### ISSUE TITLE
Emqx v3.0.1 letsencrypt not working

### ISSUE THREAD

**Issue by htaoufikallah** [emqx/emqx#2306]
#### Environment

- OS: Ubuntu 18.04
- Erlang/OTP: 
- EMQ: 3.0.1

#### Description

I followed this [medium article](https://medium.com/@emqtt/using-lets-encrypt-certificates-in-emq-b11e0e57efa6) to configure ssl using letsencrypt certificate but no success.
I reverted back to emq 2.2.0 and it works just fine.
And the dashboard also works on https and the websocket with ssl also works

- dashboard over https works  
- websocket with ssl works
- mqtt with ssl on port 8883 works

here is the test using mosquitto_sub:
mosquitto_sub -t abc -h broker-domain.com -p 8883 -d  --cafile ~/ca.pem 
Client mosqsub|21502-ubuntu-s- sending CONNECT
Client mosqsub|21502-ubuntu-s- received CONNACK
Client mosqsub|21502-ubuntu-s- sending SUBSCRIBE (Mid: 1, Topic: abc, QoS: 0)
Client mosqsub|21502-ubuntu-s- received SUBACK
Subscribed (mid: 1): 0

What am I missing to make it works for emqx v3.0.1? is there a specific guide for emqx v3.0.1 to configure ssl with letsencrypt certificate?

**Comment by terry-xiaoyu** [comment]
@htaoufikallah What is the error you got from the TLS handshake process?

**Comment by htaoufikallah** [comment]
@terry-xiaoyu Here is the error:

`mosquitto_sub -t abc -h broker.domain.com -p 8883 -d  --cafile ~/ca.pem`
Unable to connect (A TLS error occurred.).


`# ss -tln`
```
State                Recv-Q                Send-Q                                  Local Address:Port                                  Peer Address:Port                
LISTEN               0                     128                                         127.0.0.1:11883                                      0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:8080                                       0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:4369                                       0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:8883                                       0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:8083                                       0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:8084                                       0.0.0.0:*                   
LISTEN               0                     128                                     127.0.0.53%lo:53                                         0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:22                                         0.0.0.0:*                   
LISTEN               0                     5                                             0.0.0.0:5369                                       0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:1883                                       0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:6369                                       0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:18083                                      0.0.0.0:*                   
LISTEN               0                     128                                           0.0.0.0:18084                                      0.0.0.0:*                   
LISTEN               0                     128                                              [::]:4369                                          [::]:*                   
LISTEN               0                     128                                              [::]:22                                            [::]:*                   
```

```
# nc -v [server_ip] 18084
Connection to [server_ip] 18084 port [tcp/*] succeeded!

```

`# openssl s_client -connect [server_ip]:18084 -servername broker.domain.com`
```
CONNECTED(00000003)
depth=0 CN = broker.domain.com
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = broker.domain.com
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:/CN=broker.domain.com
   i:/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
---
Server certificate
-----BEGIN CERTIFICATE-----
[my cert content]
-----END CERTIFICATE-----
subject=/CN=broker.domain.com
issuer=/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
---
No client certificate CA names sent
Peer signing digest: SHA512
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 1879 bytes and written 328 bytes
Verification error: unable to verify the first certificate
---
New, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: 6EAB4EDE829EF4B96EB1F5811CD31BA9C19668D63413C3D4D29910C2365B6A6C
    Session-ID-ctx: 
    Master-Key: 5B2DEAAD399769F9AC80BBEA4F561936F537961441BB2F3A8F28A644AF9F594E27E5F45021EFEB17DFD8EF7B101433E4
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1553508619
    Timeout   : 7200 (sec)
    Verify return code: 21 (unable to verify the first certificate)
    Extended master secret: no
---
```

**Comment by phanimahesh** [comment]
Can you show relevant ssl config you did for emqx?

**Comment by htaoufikallah** [comment]
by following medium article I modified:
~/emqx/etc/plugins/emqx_dashboard.conf
```
## HTTPS Listener

## The IP address and port that the Dashboard HTTPS listener will bind.
##
## Value: IP:Port | Port
##
## Examples: 18084, 127.0.0.1:18084, ::1:18084
dashboard.listener.https = 18084

## The acceptor pool for external Dashboard HTTPS listener.
##
## Value: Number
dashboard.listener.https.acceptors = 2

## Maximum number of concurrent Dashboard HTTPS connections.
##
## Value: Number
dashboard.listener.https.max_clients = 512

## Path to the file containing the user's private PEM-encoded key.
##
## Value: File
## dashboard.listener.https.keyfile = etc/certs/key.pem
dashboard.listener.https.keyfile = /etc/letsencrypt/live/brokerdomain.com/privkey.pem

## Path to a file containing the user certificate.
##
## Value: File
## dashboard.listener.https.certfile = etc/certs/cert.pem
dashboard.listener.https.certfile = /etc/letsencrypt/live/brokerdomain.com/fullchain.pem
```
and ~/emqx/etc/emqx.conf
```
## Path to the file containing the user's private PEM-encoded key.
##
## See: listener.ssl.$name.keyfile
##
## Value: File
#listener.wss.external.keyfile = etc/certs/key.pem
listener.wss.external.keyfile = /etc/letsencrypt/live/brokerdomain.com/privkey.pem
## Path to a file containing the user certificate.
##
## See: listener.ssl.$name.certfile
##
## Value: File
#listener.wss.external.certfile = etc/certs/cert.pem
listener.wss.external.certfile = /etc/letsencrypt/live/brokerdomain.com/fullchain.pem
```

**Comment by phanimahesh** [comment]
Please also set `cacertfile` to `chain.pem`
You may also set `certfile` to `cert.pem` instead of `fullchain.pem`

For example:
```
dashboard.listener.https.keyfile = /etc/letsencrypt/live/brokerdomain.com/privkey.pem
dashboard.listener.https.certfile = /etc/letsencrypt/live/brokerdomain.com/cert.pem
dashboard.listener.https.cacertfile = /etc/letsencrypt/live/brokerdomain.com/chain.pem
```

And similarly for the wss listener. That should work.

**Comment by htaoufikallah** [comment]
Thanks @phanimahesh it worked like a charm, I think that this solution should be documented for other users.

**Comment by xerosanyam** [comment]
none of the above solution works for me
still getting
![image](https://user-images.githubusercontent.com/3167278/57175527-b7c15400-6e6a-11e9-8794-5528fbe727c8.png)

**Comment by notjulian** [comment]
@xerosanyam 
I'm using this config and it's working correctly

```

listener.ssl.external.keyfile = /etc/emqx/certs/privkey1.pem
listener.ssl.external.certfile = /etc/emqx/certs/fullchain1.pem
listener.ssl.external.cacertfile = /etc/emqx/certs/chain1.pem

listener.wss.external.keyfile = /etc/emqx/certs/privkey1.pem
listener.wss.external.certfile = /etc/emqx/certs/fullchain1.pem
listener.wss.external.cacertfile = /etc/emqx/certs/chain1.pem
```

**Comment by morganwu** [comment]
For anyone who still has this issue, you can try to change the cert file mode, to make it accessable for emqx user. This works for me.

**Comment by joko2020** [comment]
> For anyone who still has this issue, you can try to change the cert file mode, to make it accessable for emqx user. This works for me.

You mean I should do `sudo chmod` for directory path of CA file ? , if Yes What mode you use?, because I check on /etc/letsencrypt/live/mydomain it just symbolic link with access mode lrwxrwxrwx

**Comment by miguelalonso** [comment]
thanks. Solved for me, it is a question of chow the ssl certificates to the user emqx. 
In my case I used vesta control panel and after i have installed emqx

cd  /home/admin/conf/web/
chown emqx:emqx ssl.domain_name.key
chown emqx:emqx ssl.domain_name.crt
emqx restart

after taht the wss works, it can be tested in the emqx dashboard

where domain_name is your domain name. both files ssl.domain_name.key and ssl.domain_name.crt are in the /home/admin/conf/web/ folder. The files were created by letsencrypt, configured en vesta cpanel.
